### ISSUE TITLE
[ip] Bind Should Accept Zero as a Wildcard Ephemeral Port

### ISSUE THREAD

**Issue by BrianZill** [demikernel/inetstack#169]
Description
========

Currently, `bind()` will ~~allow binding to an endpoint with zero (0) as the port number part.  This is wrong, as zero is not a valid user port number~~ not allow zero (the wildcard indicator) as the port parameter, since our version of bind() takes an IPv4Endpoint which has the port field defined as a `Port16` which cannot be zero.

Expected Behavior
============

The value of zero, when provided as a port number to bind, is meant to indicate that the system should choose the port number to use from the unused ephemeral port pool.

Note this issue affects both TCP and UDP.

Related Issues
==========

Issues #168 and #108.

**Comment by ppenna** [comment]
@BrianZill I cannot confirm this issue.

1. Bind takes a `Ipv4Endpoint` structure as a parameter:

https://github.com/demikernel/inetstack/blob/e241004b7c1087ed7adc2cee7f947e0da1708fa0/src/lib.rs#L179-L190

2. `Ipv4Endpoint` is composed by a `Port16`:

https://github.com/demikernel/inetstack/blob/aed5eaa396ae7ac2f2269e1d91e5bd6406710b52/src/protocols/ipv4/endpoint.rs#L19-L24

3. `Port16` cannot be zero:

https://github.com/demikernel/runtime/blob/cb7c2be3ef5d24c8e0d37b74c7e98624dd65c41b/src/network/types/portnum.rs#L21

**Comment by BrianZill** [comment]
> @BrianZill the user cannot bind to port number zero.

Good catch.  I fixed the Issue Description.

> 3. `Port16` cannot be zero:

Another bug.  Not sure if we should create a separate Issue for this, or just fix it when fixing this one.

Technically, outside of this traditional API usage of 0 as a wildcard for bind, I don't think that there is anything in the TCP spec itself that prevents the usage of port 0.  As far as the TCP spec goes, 0 is just another port number.  The IANA has port 0 marked as "Reserved".

UDP, however, actually specifies zero as a valid source port value that basically means "unspecified port".  So, it is explicitly allowed to appear in a UDP packet header.  I just looked, and our implementation represents the UDP header's source port field as an `Option<Port16>`.  This looks like another case of someone who "never met a language feature they didn't like".  I suspect `Port16` is defined as non-zero just so this line in the UDP parser could be written this way:

`let src_port: Option<Port16> = Port16::try_from(NetworkEndian::read_u16(&hdr_buf[0..2])).ok();`

Which, while cute, ignores the fact that there are plenty of other places where port numbers are allowed to be zero.  It's also wasteful, as the UDP header's serialize routine now needs to know that if the "optional" source port isn't there, it should write zeroes to that field in the header.

I've created Issue #172 to track this.
